pkg/storage/fs/ocis/ocis.go

```diff
@@ -183,8 +189,55 @@ func (fs *ocisfs) Shutdown(ctx context.Context) error {
 	return nil
 }
 
-func (fs *ocisfs) GetQuota(ctx context.Context) (int, int, error) {
-	return 0, 0, nil
+// TODO Document in the cs3 should we return quota or free space?
+func (fs *ocisfs) GetQuota(ctx context.Context) (uint64, uint64, error) {
+	var node *Node
+	var err error
+	if node, err = fs.lu.HomeOrRootNode(ctx); err != nil {
+		return 0, 0, err
+	}
+
+	if !node.Exists {
+		err = errtypes.NotFound(filepath.Join(node.ParentID, node.Name))
+		return 0, 0, err
+	}
+
+	rp, err := fs.p.AssemblePermissions(ctx, node)
+	switch {
+	case err != nil:
+		return 0, 0, errtypes.InternalError(err.Error())
+	case !rp.GetQuota:
+		return 0, 0, errtypes.PermissionDenied(node.ID)
+	}
+
+	ri, err := node.AsResourceInfo(ctx, rp, []string{"treesize", "quota"})
+	if err != nil {
+		return 0, 0, err
+	}
+
+	quotaStr := _quotaUnknown
+	if ri.Opaque != nil && ri.Opaque.Map != nil && ri.Opaque.Map["quota"] != nil && ri.Opaque.Map["quota"].Decoder == "plain" {
+		quotaStr = string(ri.Opaque.Map["quota"].Value)
+	}
+	stat := syscall.Statfs_t{}
+	err = syscall.Statfs(fs.lu.toInternalPath(node.ID), &stat)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	total := stat.Blocks * uint64(stat.Bsize) // Total data blocks in filesystem
```

Comment:
Shouldn't `total` be the available disk space? 🤔 at least from a user perspective it's misleading that with unlimited quota the response will tell that (on my machine) 495GB are free, while my free disk space is at 195GB.

Reply:
uh ... that is weird. maybe the total is calculated differently on mac?

Reply:
macOS is weird for `stat`... being in the ocis user home dir of einstein:

```
> stat -x .
  File: "."
  Size: 128          FileType: Directory
  Mode: (0700/drwx------)         Uid: (  501/    bene)  Gid: (    0/   wheel)
Device: 1,5   Inode: 62850061    Links: 4
Access: Wed Feb 10 07:33:06 2021
Modify: Wed Feb 10 07:33:06 2021
Change: Wed Feb 10 09:56:26 2021
```

```
> stat .
16777221 62850061 drwx------ 4 bene wheel 0 128 "Feb 10 07:33:06 2021" "Feb 10 07:33:06 2021" "Feb 10 09:56:26 2021" "Feb 10 07:23:36 2021" 4096 0 0 .
```

Reply:
ok, what about filesystem stat:
```
> stat -f .
  File: "."
    ID: fedc9aa3bd65bc57 Namelen: 255     Type: ext2/ext3
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 65793553   Free: 58593039   Available: 55233500
Inodes: Total: 16777216   Free: 16011441
```

Reply:
macOS doesn't have `stat` for the entire filesystem AFAICT.
```
> BLOCKSIZE=4K df
Filesystem     4K-blocks     Used Available Capacity iused      ifree %iused  Mounted on
/dev/disk1s5s1 122061322  4756743  34314966    13%  568975 4881883905    0%   /
devfs                 47       47         0   100%     658          0  100%   /dev
/dev/disk1s4   122061322   524298  34314966     2%       2 4882452878    0%   /System/Volumes/VM
/dev/disk1s2   122061322    80331  34314966     1%     792 4882452088    0%   /System/Volumes/Preboot
/dev/disk1s6   122061322    27269  34314966     1%     396 4882452484    0%   /System/Volumes/Update
/dev/disk1s1   122061322 82168442  34314966    71% 4246575 4878206305    0%   /System/Volumes/Data
map auto_home          0        0         0   100%       0          0  100%   /System/Volumes/Data/home
/dev/disk1s5   122061322  4756743  34314966    13%  568977 4881883903    0%   /System/Volumes/Update/mnt1
```

Reply:
ah now I get it ... the switch case in https://github.com/cs3org/reva/pull/1405/files#diff-62fe53c34a46cf0bf6e8f27409f0ba1ba6b0687387b8c9cde985299ba10b839fR229-R240 should limit the total to the quota  if the latter is set. the free space is then calculated in the ocs / ocdav handler by calculating `free = total-used`. The problem is that we currently cannot return 'unlimited' quota. It would be a dedicated flag because currently quota is an unsigned int...

Reply:
actually, if no quote is set we should return free+used as the total, because other users files also count against the total.